File: tools/publish/publish.py
"""Incremental publisher; Python 3.10+, standard library only."""
import argparse
import base64
import hashlib
import json
import os
from pathlib import Path
import re
import sys
import urllib.request
import urllib.parse

EXTENSIONS = {'html', 'css', 'js', 'png', 'gif', 'jpg', 'jpeg', 'svg', 'ico', 'cur', 'webp', 'psd', 'sym', 'txt', 'pdf', 'woff', 'woff2'}


def valid_path(path):
    return (len(path) <= 512 and all(
        re.fullmatch(r'[A-Za-z0-9_-][A-Za-z0-9_. ()-]*', p)
        and not p.endswith(('.', ' '))
        and not re.search(r'\.(php[0-9]*|phtml|phar|cgi|pl|py|sh|shtml)(\.|$)', p, re.I)
        for p in path.split('/')) and path.rsplit('.', 1)[-1].lower() in EXTENSIONS)


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        raise RuntimeError('Endpoint redirected; use its exact HTTPS URL')


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--endpoint', default=os.environ.get('FODOCS_PUBLISH_URL'))
    parser.add_argument('--site', type=Path, default=Path(__file__).resolve().parents[2] / '_site')
    parser.add_argument('--apply', action='store_true', help='Perform changes (default: dry run)')
    parser.add_argument('--delete', action='store_true', help='Also remove remote static files absent locally')
    args = parser.parse_args()
    url = urllib.parse.urlsplit(args.endpoint or '')
    if url.scheme != 'https' or not url.hostname or url.username or url.password or url.query or url.fragment:
        parser.error('Provide an exact HTTPS endpoint without credentials, query, or fragment')
    key = os.environ.get('FODOCS_PUBLISH_KEY', '')
    if len(key) < 64:
        parser.error('Set FODOCS_PUBLISH_KEY to the shared random key (at least 64 characters)')
    if args.site.is_symlink() or not (args.site / 'index.html').is_file():
        parser.error('Site must be a real directory containing index.html; build first')
    local = {}
    for root, dirs, files in os.walk(args.site, followlinks=False):
        for name in dirs + files:
            if (Path(root) / name).is_symlink():
                raise RuntimeError('Symlinks are forbidden in the output site')
        for name in files:
            file = Path(root) / name
            relative = file.relative_to(args.site).as_posix()
            if not valid_path(relative):
                raise RuntimeError(f'Unsupported static path: {relative}')
            data = file.read_bytes()
            if len(data) > 24 * 1024 * 1024 - 4096:
                raise RuntimeError(f'File exceeds endpoint request limit: {relative}')
            local[relative] = (data, hashlib.sha256(data).hexdigest())
    opener = urllib.request.build_opener(NoRedirect())

    def request(payload):
        req = urllib.request.Request(args.endpoint, json.dumps(payload).encode(), {
            'Authorization': 'Bearer ' + key, 'Content-Type': 'application/json'})
        try:
            with opener.open(req, timeout=120) as response:
                return json.load(response)
        except urllib.error.HTTPError as error:
            raise RuntimeError(
                f"{payload['op']} {payload.get('path', '')}: HTTP {error.code}; "
                'check the VPS Apache/PHP error log'
            ) from error

    remote = request({'op': 'list'})['files']
    if not isinstance(remote, dict) or any(not valid_path(p) for p in remote):
        raise RuntimeError('Invalid server manifest')
    puts = sorted(p for p, (_, h) in local.items() if remote.get(p, {}).get('sha256') != h)
    deletes = sorted(set(remote) - set(local)) if args.delete else []
    for p in puts:
        print('PLAN PUT   ', p)
    for p in deletes:
        print('PLAN DELETE', p)
    print(f'{len(puts)} uploads, {len(deletes)} deletions' + ('' if args.apply else ' (dry run)'))
    if not args.apply:
        return
    # Upload everything successfully before attempting deletions.
    for p in puts:
        data, digest = local[p]
        request({'op': 'put', 'path': p, 'expected': remote.get(p, {}).get('sha256'),
                 'sha256': digest, 'data': base64.b64encode(data).decode('ascii')})
        print('Uploaded', p, flush=True)
    for p in deletes:
        request({'op': 'delete', 'path': p, 'expected': remote[p]['sha256']})
        print('Deleted', p, flush=True)
    verified = request({'op': 'list'})['files']
    if any(verified.get(p, {}).get('sha256') != h for p, (_, h) in local.items()) or any(p in verified for p in deletes):
        raise RuntimeError('Post-publish verification failed; rerun to reconcile')
    print('Published and verified.')


if __name__ == '__main__':
    try:
        main()
    except Exception as error:
        print(f'Publish failed: {error}', file=sys.stderr)
        sys.exit(1)

File: tools/publish/publish.php
<?php
declare(strict_types=1);

// Install in the site root. Configure these through PHP-FPM / Apache, not HTTP.
// FODOCS_PUBLISH_KEY: at least 32 random bytes encoded as hex.
// FODOCS_PUBLISH_IPS: optional comma-separated REMOTE_ADDR allowlist.
ini_set('display_errors', '0');
header('Content-Type: application/json');
header('Cache-Control: no-store');
const MAX_BYTES = 33554432;
const EXTENSIONS = ['html', 'css', 'js', 'png', 'gif', 'jpg', 'jpeg', 'svg', 'ico', 'cur', 'webp', 'psd', 'sym', 'txt', 'pdf', 'woff', 'woff2'];

function reply(int $status, array $data): never {
    http_response_code($status);
    echo json_encode($data, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES);
    exit;
}
function validPath(string $path): bool {
    if (strlen($path) > 512 || !preg_match('~\A[A-Za-z0-9_-][A-Za-z0-9_. /()-]*\z~D', $path)) return false;
    foreach (explode('/', $path) as $part) {
        if (!preg_match('~\A[A-Za-z0-9_-][A-Za-z0-9_. ()-]*\z~D', $part) || str_ends_with($part, '.') || str_ends_with($part, ' ')) return false;
        // Reject compound script names even if the final extension is static.
        if (preg_match('~\.(php[0-9]*|phtml|phar|cgi|pl|py|sh|shtml)(\.|$)~i', $part)) return false;
    }
    return in_array(strtolower(pathinfo($path, PATHINFO_EXTENSION)), EXTENSIONS, true);
}
function target(string $path, bool $create): string {
    if (!validPath($path)) reply(400, ['error' => 'Invalid static file path']);
    $parts = explode('/', $path);
    $current = __DIR__;
    foreach ($parts as $i => $part) {
        $current .= '/' . $part;
        clearstatcache(true, $current);
        if (is_link($current)) reply(400, ['error' => 'Symlinks are forbidden']);
        if ($i < count($parts) - 1) {
            if (!file_exists($current) && $create && !mkdir($current, 0755)) throw new RuntimeException('mkdir failed');
            if (file_exists($current) && !is_dir($current)) reply(409, ['error' => 'Parent is not a directory']);
        } elseif (file_exists($current) && !is_file($current)) reply(409, ['error' => 'Target is not a regular file']);
    }
    return $current;
}

try {
    if (($_SERVER['HTTPS'] ?? '') !== 'on' && ($_SERVER['HTTPS'] ?? '') !== '1') reply(403, ['error' => 'HTTPS required']);
    $key = getenv('FODOCS_PUBLISH_KEY');
    if (!$key || strlen($key) < 64) reply(503, ['error' => 'Publishing is not configured']);
    $ips = array_filter(array_map('trim', explode(',', getenv('FODOCS_PUBLISH_IPS') ?: '')));
    if ($ips && !in_array($_SERVER['REMOTE_ADDR'] ?? '', $ips, true)) reply(403, ['error' => 'Forbidden']);
    if (!hash_equals('Bearer ' . $key, $_SERVER['HTTP_AUTHORIZATION'] ?? '')) reply(401, ['error' => 'Unauthorized']);
    if ($_SERVER['REQUEST_METHOD'] !== 'POST') reply(405, ['error' => 'POST required']);
    // Lock the endpoint itself; no writable state files in the web root.
    $lock = fopen(__FILE__, 'r');
    if (!$lock || !flock($lock, LOCK_EX)) throw new RuntimeException('Lock failed');
    $body = file_get_contents('php://input', false, null, 0, MAX_BYTES + 1);
    if ($body === false || strlen($body) > MAX_BYTES) reply(413, ['error' => 'Request too large']);
    try { $request = json_decode($body, true, 16, JSON_THROW_ON_ERROR); }
    catch (JsonException $e) { reply(400, ['error' => 'Invalid JSON']); }
    if (!is_array($request)) reply(400, ['error' => 'Expected object']);
    $op = $request['op'] ?? '';
    if ($op === 'list') {
        $files = [];
        $dirs = [];
        $tree = new RecursiveCallbackFilterIterator(
            new RecursiveDirectoryIterator(__DIR__, FilesystemIterator::SKIP_DOTS),
            function ($entry): bool {
                if ($entry->isLink()) return false;
                $relative = str_replace(DIRECTORY_SEPARATOR, '/', substr($entry->getPathname(), strlen(__DIR__) + 1));
                return validPath($entry->isDir() ? $relative . '/file.html' : $relative);
            }
        );
        $iterator = new RecursiveIteratorIterator($tree, RecursiveIteratorIterator::SELF_FIRST);
        foreach ($iterator as $entry) {
            if ($entry->isLink()) continue;
            $relative = str_replace(DIRECTORY_SEPARATOR, '/', substr($entry->getPathname(), strlen(__DIR__) + 1));
            if ($entry->isDir()) { $dirs[] = $relative; continue; }
            if (!$entry->isFile() || !validPath($relative)) continue;
            $hash = hash_file('sha256', $entry->getPathname());
            if ($hash === false) throw new RuntimeException('Hash failed');
            $files[$relative] = ['sha256' => $hash, 'size' => $entry->getSize()];
        }
        ksort($files);
        sort($dirs);
        reply(200, ['files' => (object)$files, 'directories' => $dirs]);
    }
    if (!in_array($op, ['put', 'delete'], true) || !is_string($request['path'] ?? null)) reply(400, ['error' => 'Invalid operation']);
    $path = target($request['path'], false);
    $currentHash = is_file($path) ? hash_file('sha256', $path) : null;
    if (!array_key_exists('expected', $request) || $request['expected'] !== $currentHash) reply(409, ['error' => 'File changed; list again']);
    if ($op === 'delete') {
        if (is_file($path) && !unlink($path)) throw new RuntimeException('Delete failed');
        reply(200, ['ok' => true]);
    }
    $data = is_string($request['data'] ?? null) ? base64_decode($request['data'], true) : false;
    if ($data === false || !is_string($request['sha256'] ?? null) || !hash_equals(hash('sha256', $data), $request['sha256'])) reply(400, ['error' => 'Invalid data or hash']);
    $path = target($request['path'], true);
    $temp = tempnam(dirname($path), '.publish-');
    if ($temp === false) throw new RuntimeException('Temp failed');
    try {
        if (file_put_contents($temp, $data) !== strlen($data) || !chmod($temp, 0644) || !rename($temp, $path)) throw new RuntimeException('Write failed');
    } finally { if (is_file($temp)) unlink($temp); }
    reply(200, ['ok' => true]);
} catch (Throwable $e) {
    error_log('FoDocs publish: ' . $e->getMessage());
    reply(500, ['error' => 'Publishing failed; check server log']);
}
